linera-rpc/src/grpc/server.rs

```diff
@@ -399,75 +401,133 @@ where
         cross_chain_sender_failure_rate: f32,
         cross_chain_max_concurrent_tasks: usize,
         this_shard: ShardId,
-        receiver: mpsc::Receiver<(linera_core::data_types::CrossChainRequest, ShardId)>,
+        mut receiver: mpsc::Receiver<(linera_core::data_types::CrossChainRequest, ShardId)>,
     ) {
         let pool = GrpcConnectionPool::default();
-        let max_concurrent_tasks = Some(cross_chain_max_concurrent_tasks);
-
-        receiver
-            .for_each_concurrent(max_concurrent_tasks, |(cross_chain_request, shard_id)| {
-                let shard = network.shard(shard_id);
-                let remote_address = shard.http_address();
-
-                let pool = pool.clone();
-                let nickname = nickname.clone();
-
+        let mut join_set = JoinSet::new();
+        // An entry means that a task is running for the given sender, recipient and request kind.
+        // If it is Some, that means the task is still working on another request, and will pick
+        // up the new one when it is done. Since later requests always supersede earlier ones,
+        // we can always overwrite the entry if a new one comes in.
+        let cross_chain_tasks = Arc::new(Mutex::new(HashMap::new()));
+        let semaphore = Arc::new(tokio::sync::Semaphore::new(
+            cross_chain_max_concurrent_tasks,
```

Comment:
Maybe we don't need `cross_chain_max_concurrent_tasks` any more

Reply:
I'm not sure: Since receiver chains don't need to exist, users can create an unlimited number of tasks that way.

On the other hand that's kind of already the case, since we don't count sleeping tasks towards the limit.

Reply:
Let's keep discussing this